src/python/pants/backend/python/goals/package_pex_binary.py

```diff
@@ -101,11 +106,33 @@ def generate_additional_args(self, pex_binary_defaults: PexBinaryDefaults) -> Tu
 
 @rule(level=LogLevel.DEBUG)
 async def package_pex_binary(
-    field_set: PexBinaryFieldSet, pex_binary_defaults: PexBinaryDefaults
+    field_set: PexBinaryFieldSet,
+    pex_binary_defaults: PexBinaryDefaults,
+    union_membership: UnionMembership,
 ) -> BuiltPackage:
-    resolved_entry_point = await Get(
-        ResolvedPexEntryPoint, ResolvePexEntryPointRequest(field_set.entry_point)
+    resolved_entry_point, transitive_targets = await MultiGet(
+        Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest(field_set.entry_point)),
+        Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address])),
     )
+
+    # If a zip app, check for files targets because these will not be loadable as expected.
+    if field_set._execution_mode == PexExecutionMode.ZIPAPP:
```

Comment:
According to @jsirois' comment, this should check `PexExecutionMode.ZIPAPP` _and_ `zip_safe=True`, I think.